### DISCUSSION TITLE
How to determine which module is adding bootstrap 5 into my app?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#10113] [Q&A] [answered]
I recently upgraded all my packaged to `1.1.0-preview-16417`. After the upgrade, something is including bootstrap 5 into my site where my custom there already injecting bootstrap 4. I need to find out which module is using bootstrap 5 and stop it from injecting it.

In my custom theme, I have added the following 

```
    <script asp-name="bootstrap" version="4" at="Foot"></script>
    <style asp-name="CrestAppsBootstrap4" version="1"></style> <!-- this file include bootstrap 4.6 css + some overrides that I need -->
```

However, the final output seems to bootstrap 5 and bootstrap 4

Here is what is being generated into the head section
  ```
<!-- This is bootstrap 5 which is unwanted -->  
<link href="/OrchardCore.Resources/Styles/bootstrap.css?v=mQbsKz3KorwjzMw3dQNuZ5k7lMNBFi6q1bGbyA6Cqjk" rel="stylesheet" type="text/css" />

<!-- this is my custom bootstrap 4 -->
<link href="/CrestApps.Themes.Bootstrap4/css/custom-bootstrap.css?v=W_GvZKSQn4NwUZAB4bLtKNqLVyT3L0V075IIAYypDkg" rel="stylesheet" type="text/css" />
```

Here is what is being generated in the foot

```
<script src="/OrchardCore.Resources/Scripts/jquery.js?v=a7Mta-Fwf7X9I3_MTEW_iuYCpv7pXJScHQTJuri3ewM"></script>
<script src="/OrchardCore.Resources/Scripts/popper.js?v=7dZ18llGAY2fVQ5Q3Bod-UoRtGiwtOqmadaDpvy4W2Y"></script>

<!-- This is bootstrap 5 which is unwanted -->  
<script src="/OrchardCore.Resources/Scripts/bootstrap.js?v=JrvqF4l5dn5tI4I-j6PooTxvDVg9atOfOmIPkR8kDEQ"></script>

<script src="/OrchardCore.Resources/Vendor/popper-1.16.1/popper.js?v=idFXvDm9Bjff32pnnuNRfJguUWUX97DFijva9TGScPs"></script>
<script src="/OrchardCore.Resources/Vendor/bootstrap-4.6.0/bootstrap.js?v=J0gMaYMDNMez_iRLZyV9S9Kmd4l4ZC5bArhNgbfTmOo"></script>
```

**Comment by Piedone** [comment, marked as answer]
Isn't your theme using a built-in theme as its base theme by chance? Because that may include Bootstrap too. I don't think anything else included Boostrap on the frontend.

Also, if in your `ResourceManagementOptionsConfiguration` classes you have `SetDependencies("bootstrap")` anywhere that'll include the latest version, i.e. 5.1 currently, too.

I'm not sure about the `version` attribute though, probably it's a mismatch and you should use the string `"4.6.0"` (since that's the specific version).

**Reply by MikeAlhayek** [reply]
@Piedone Thank you for that info! My custom these does not depend on any other these. I think I copied `TheBlogTheme` and changed it.

I was using `SetDependencies("bootstrap")` in one of my resource. but I though that would utilize the bootstrap version that is being used not automatically add the latest version. There must be a way to set the dependencies without creating duplicated. Isn't there?

Any way, I removed  `SetDependencies("bootstrap")` from the configuration-options class. However, I am still having the same issue. Is it possible that the dependencies are cache? If so, how do I clear it? Per the [document ](https://docs.orchardcore.net/en/dev/docs/reference/modules/Resources/) using `version="4"` should use the latest version of bootstrap 4. Any way I tried specifying "4.6.0" instead of "4" and got the same result.

> This example will use the latest available version with a Major version of 3, like 3.4.0. If the version is not specified the latest one is always used.

**Reply by MikeAlhayek** [reply]
Ahhh stupid browser cache! Ctrl + F5 did the trick! When setting the dependencies, is there a way to set the dependencies for included version or the latest? so in this case `SetDependencies("bootstrap")` would not generated bootstrap 5.1 since 4.6 is already included?

**Reply by Piedone** [reply]
I don't think you can define dependency versions, unfortunately. Could you open a feature request, please?

**Reply by MikeAlhayek** [reply]
@Piedone I opened a feature request as per your request #10117

**Reply by Piedone** [reply]
Thanks!
